File: Back-End-System/src/services/userService.js
const bcrypt = require('bcrypt');
const prisma = require('../prisma');
const { pool } = require('../db');
const { toDbRole, toApiRole } = require('../utils/mappers');
const authSecurityService = require('./authSecurityService');

const toPublicUser = (row) => ({
  user_id: Number(row.userId),
  first_name: row.firstName,
  last_name: row.lastName,
  email: row.email,
  role: toApiRole(row.role),
  is_active: row.isActive,
  created_at: row.createdAt,
  updated_at: row.updatedAt,
});

const listUsers = async ({ search, page, limit }) => {
  const p = Math.max(Number(page) || 1, 1);
  const l = Math.max(Number(limit) || 10, 1);

  const where = search
    ? {
        OR: [
          { firstName: { contains: String(search), mode: 'insensitive' } },
          { lastName: { contains: String(search), mode: 'insensitive' } },
          { email: { contains: String(search), mode: 'insensitive' } },
          { role: { contains: String(search), mode: 'insensitive' } },
        ],
      }
    : {};

  const users = await prisma.user.findMany({
    where,
    orderBy: { userId: 'desc' },
    skip: page || limit ? (p - 1) * l : undefined,
    take: page || limit ? l : undefined,
  });

  return users.map(toPublicUser);
};

const getUserById = async (userId) => {
  const user = await prisma.user.findUnique({ where: { userId: Number(userId) } });
  return user ? toPublicUser(user) : null;
};

const getUserStats = async () => {
  const [totalUsers, byRole] = await Promise.all([
    prisma.user.count(),
    prisma.user.groupBy({ by: ['role'], _count: { role: true } }),
  ]);

  const roles = {};
  byRole.forEach((row) => {
    roles[toApiRole(row.role)] = Number(row._count.role || 0);
  });

  return { total_users: Number(totalUsers), roles };
};

const createUser = async (payload) => {
  const hashed = await bcrypt.hash(String(payload.password), 10);
  const role = toDbRole(payload.role);

  const user = await prisma.$transaction(async (tx) => {
    const created = await tx.user.create({
      data: {
        email: String(payload.email).trim().toLowerCase(),
        passwordHash: hashed,
        role,
        firstName: String(payload.first_name).trim(),
        lastName: String(payload.last_name).trim(),
        isActive: true,
      },
    });

    if (role === 'student') {
      const studentNumber = `STU${String(created.userId).padStart(6, '0')}`;
      await tx.$executeRawUnsafe(
        `INSERT INTO students(user_id,student_number,classification,major_id,admission_date,credits_earned,gpa,academic_standing,advisor_id)
         VALUES($1,$2,$3,$4,$5,$6,$7,$8,$9)`,
        Number(created.userId),
        studentNumber,
        'freshman',
        null,
        null,
        0,
        0,
        'good',
        null
      );
    }

    return created;
  });

  await authSecurityService.setMustChangePassword(user.userId, true);
  return toPublicUser(user);
};

const updateUser = async (userId, payload) => {
  const id = Number(userId);
  const existing = await prisma.user.findUnique({ where: { userId: id } });
  if (!existing) return null;

  const data = {
    email: String(payload.email).trim().toLowerCase(),
    firstName: String(payload.first_name).trim(),
    lastName: String(payload.last_name).trim(),
    role: toDbRole(payload.role),
    updatedAt: new Date(),
  };

  if (payload.password) {
    data.passwordHash = await bcrypt.hash(String(payload.password), 10);
  }

  const updated = await prisma.user.update({
    where: { userId: id },
    data,
  });

  return toPublicUser(updated);
};

const patchUser = async (userId, payload) => {
  const id = Number(userId);
  const existing = await prisma.user.findUnique({ where: { userId: id } });
  if (!existing) return null;

  const data = {
    updatedAt: new Date(),
  };

  if (payload.email) data.email = String(payload.email).trim().toLowerCase();
  if (payload.first_name !== undefined) data.firstName = String(payload.first_name);
  if (payload.last_name !== undefined) data.lastName = String(payload.last_name);
  if (payload.role) data.role = toDbRole(payload.role);
  if (payload.is_active !== undefined) data.isActive = Boolean(payload.is_active);
  if (payload.password) data.passwordHash = await bcrypt.hash(String(payload.password), 10);

  const updated = await prisma.user.update({
    where: { userId: id },
    data,
  });

  return toPublicUser(updated);
};

const deleteUser = async (userId) => {
  const id = Number(userId);
  if (!id) return false;

  const client = await pool.connect();
  try {
    await client.query('BEGIN');

    const existing = await client.query('SELECT user_id FROM users WHERE user_id=$1 FOR UPDATE', [id]);
    if (!existing.rowCount) {
      await client.query('ROLLBACK');
      return false;
    }

    // If this user owns student records, delete dependent academic rows first.
    const students = await client.query('SELECT student_id FROM students WHERE user_id=$1', [id]);
    for (const row of students.rows) {
      const studentId = Number(row.student_id);
      await client.query('DELETE FROM grades WHERE enrollment_id IN (SELECT enrollment_id FROM enrollments WHERE student_id=$1)', [studentId]);
      await client.query('DELETE FROM finance_records WHERE student_id=$1', [studentId]);
      await client.query('DELETE FROM pre_registered_courses WHERE student_id=$1', [studentId]);
      await client.query('DELETE FROM enrollments WHERE student_id=$1', [studentId]);
      await client.query('DELETE FROM students WHERE student_id=$1', [studentId]);
    }

    // If this user owns faculty records, detach sections before deleting faculty.
    const faculty = await client.query('SELECT faculty_id FROM faculty WHERE user_id=$1', [id]);
    for (const row of faculty.rows) {
      const facultyId = Number(row.faculty_id);
      await client.query('UPDATE students SET advisor_id=NULL WHERE advisor_id=$1', [facultyId]);
      await client.query('UPDATE class_sections SET faculty_id=NULL WHERE faculty_id=$1', [facultyId]);
      await client.query('DELETE FROM faculty WHERE faculty_id=$1', [facultyId]);
    }

    // Detach optional audit references to this user.
    await client.query('UPDATE grades SET posted_by=NULL WHERE posted_by=$1', [id]);
    await client.query('UPDATE pre_registered_courses SET approved_by=NULL WHERE approved_by=$1', [id]);

    await client.query('DELETE FROM users WHERE user_id=$1', [id]);
    await client.query('COMMIT');
    return true;
  } catch (error) {
    await client.query('ROLLBACK');
    if (error?.code === '23503') {
      const conflict = new Error('Cannot delete user because it is linked to other records');
      conflict.statusCode = 409;
      throw conflict;
    }
    throw error;
  } finally {
    client.release();
  }
};

module.exports = {
  listUsers,
  getUserById,
  getUserStats,
  createUser,
  updateUser,
  patchUser,
  deleteUser,
};

File: Testing/app/services/user_service.py
"""Mirrors Back-End/src/services/userService.js"""

from passlib.context import CryptContext
from sqlalchemy import text
from sqlalchemy.orm import Session

from app.services import auth_security_service
from app.utils.mappers import to_api_role, to_db_role

pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")


def _to_public(r) -> dict:
    return {
        "user_id": int(r["user_id"]),
        "first_name": r["first_name"],
        "last_name": r["last_name"],
        "email": r["email"],
        "role": to_api_role(r["role"]),
        "is_active": r["is_active"],
        "created_at": r["created_at"],
        "updated_at": r["updated_at"],
    }


def list_users(db: Session, search: str | None = None, page: int = 1, limit: int = 10) -> list:
    p = max(int(page or 1), 1)
    l = max(int(limit or 10), 1)
    params: dict = {"limit": l, "offset": (p - 1) * l}

    if search:
        params["search"] = f"%{search.lower()}%"
        where = (
            "WHERE LOWER(first_name) LIKE :search OR LOWER(last_name) LIKE :search "
            "OR LOWER(email) LIKE :search OR LOWER(role) LIKE :search"
        )
    else:
        where = ""

    rows = db.execute(
        text(
            f"SELECT user_id, first_name, last_name, email, role, is_active, created_at, updated_at "
            f"FROM users {where} ORDER BY user_id DESC LIMIT :limit OFFSET :offset"
        ),
        params,
    ).mappings().all()

    return [_to_public(r) for r in rows]


def get_user_by_id(db: Session, user_id: int) -> dict | None:
    row = db.execute(
        text(
            "SELECT user_id, first_name, last_name, email, role, is_active, created_at, updated_at "
            "FROM users WHERE user_id = :uid"
        ),
        {"uid": int(user_id)},
    ).mappings().first()
    return _to_public(row) if row else None


def get_user_stats(db: Session) -> dict:
    total = db.execute(text("SELECT COUNT(*) AS cnt FROM users")).mappings().first()["cnt"]
    role_rows = db.execute(
        text("SELECT role, COUNT(*) AS cnt FROM users GROUP BY role")
    ).mappings().all()
    roles = {to_api_role(r["role"]): int(r["cnt"]) for r in role_rows}
    return {"total_users": int(total), "roles": roles}


def create_user(db: Session, payload: dict) -> dict:
    hashed = pwd_context.hash(str(payload.get("password", "")))
    role = to_db_role(str(payload.get("role", "student")))

    db.execute(
        text(
            "INSERT INTO users(email, password_hash, role, first_name, last_name, is_active) "
            "VALUES (:email, :hash, :role, :first, :last, true)"
        ),
        {
            "email": str(payload.get("email", "")).strip().lower(),
            "hash": hashed,
            "role": role,
            "first": str(payload.get("first_name", "")).strip(),
            "last": str(payload.get("last_name", "")).strip(),
        },
    )

    row = db.execute(
        text("SELECT user_id FROM users WHERE LOWER(email) = LOWER(:email) LIMIT 1"),
        {"email": str(payload.get("email", "")).strip().lower()},
    ).mappings().first()

    user_id = int(row["user_id"])

    if role == "student":
        student_number = f"STU{str(user_id).zfill(6)}"
        db.execute(
            text(
                "INSERT INTO students(user_id, student_number, classification, major_id, "
                "admission_date, credits_earned, gpa, academic_standing, advisor_id) "
                "VALUES (:uid, :snum, 'freshman', NULL, NULL, 0, 0, 'good', NULL)"
            ),
            {"uid": user_id, "snum": student_number},
        )

    db.commit()
    auth_security_service.set_must_change_password(db, user_id, True)
    return get_user_by_id(db, user_id)


def update_user(db: Session, user_id: int, payload: dict) -> dict | None:
    existing = get_user_by_id(db, user_id)
    if not existing:
        return None

    params = {
        "email": str(payload.get("email", existing["email"])).strip().lower(),
        "first": str(payload.get("first_name", existing["first_name"])).strip(),
        "last": str(payload.get("last_name", existing["last_name"])).strip(),
        "role": to_db_role(str(payload.get("role", existing["role"]))),
        "uid": int(user_id),
    }

    if payload.get("password"):
        params["hash"] = pwd_context.hash(str(payload["password"]))
        db.execute(
            text(
                "UPDATE users SET email=:email, first_name=:first, last_name=:last, "
                "role=:role, password_hash=:hash, updated_at=NOW() WHERE user_id=:uid"
            ),
            params,
        )
    else:
        db.execute(
            text(
                "UPDATE users SET email=:email, first_name=:first, last_name=:last, "
                "role=:role, updated_at=NOW() WHERE user_id=:uid"
            ),
            params,
        )

    db.commit()
    return get_user_by_id(db, user_id)


def patch_user(db: Session, user_id: int, payload: dict) -> dict | None:
    existing = get_user_by_id(db, user_id)
    if not existing:
        return None

    updates = []
    params: dict = {"uid": int(user_id)}

    if payload.get("email"):
        updates.append("email = :email")
        params["email"] = str(payload["email"]).strip().lower()
    if payload.get("first_name") is not None:
        updates.append("first_name = :first_name")
        params["first_name"] = str(payload["first_name"])
    if payload.get("last_name") is not None:
        updates.append("last_name = :last_name")
        params["last_name"] = str(payload["last_name"])
    if payload.get("role"):
        updates.append("role = :role")
        params["role"] = to_db_role(str(payload["role"]))
    if payload.get("is_active") is not None:
        updates.append("is_active = :is_active")
        params["is_active"] = bool(payload["is_active"])
    if payload.get("password"):
        updates.append("password_hash = :hash")
        params["hash"] = pwd_context.hash(str(payload["password"]))

    if not updates:
        return existing

    updates.append("updated_at = NOW()")
    db.execute(
        text(f"UPDATE users SET {', '.join(updates)} WHERE user_id = :uid"),
        params,
    )
    db.commit()
    return get_user_by_id(db, user_id)


def delete_user(db: Session, user_id: int) -> bool:
    result = db.execute(
        text("DELETE FROM users WHERE user_id = :uid RETURNING user_id"),
        {"uid": int(user_id)},
    ).mappings().first()
    db.commit()
    return result is not None
